return {
	__mul = function(t,a,b)
t[1]=a[1]*b[1]+a[3]*b[3]+a[4]*b[4]-a[7]*b[7]
t[2]=a[2]*b[1]+a[1]*b[2]+a[5]*b[3]+a[6]*b[4]-a[3]*b[5]-a[4]*b[6]-a[8]*b[7]-a[7]*b[8]
t[3]=a[3]*b[1]+a[1]*b[3]+a[7]*b[4]-a[4]*b[7]
t[4]=a[4]*b[1]-a[7]*b[3]+a[1]*b[4]+a[3]*b[7]
t[5]=a[5]*b[1]-a[3]*b[2]+a[2]*b[3]+a[8]*b[4]+a[1]*b[5]+a[7]*b[6]-a[6]*b[7]+a[4]*b[8]
t[6]=a[6]*b[1]-a[4]*b[2]-a[8]*b[3]+a[2]*b[4]-a[7]*b[5]+a[1]*b[6]+a[5]*b[7]-a[3]*b[8]
t[7]=a[7]*b[1]-a[4]*b[3]+a[3]*b[4]+a[1]*b[7]
t[8]=a[8]*b[1]+a[7]*b[2]-a[6]*b[3]+a[5]*b[4]+a[4]*b[5]-a[3]*b[6]+a[2]*b[7]+a[1]*b[8]
	end,
	__pow = function(t,a,b)
t[1]=a[1]*b[1]
t[2]=a[2]*b[1]+a[1]*b[2]
t[3]=a[3]*b[1]+a[1]*b[3]
t[4]=a[4]*b[1]+a[1]*b[4]
t[5]=a[5]*b[1]-a[3]*b[2]+a[2]*b[3]+a[1]*b[5]
t[6]=a[6]*b[1]-a[4]*b[2]+a[2]*b[4]+a[1]*b[6]
t[7]=a[7]*b[1]-a[4]*b[3]+a[3]*b[4]+a[1]*b[7]
t[8]=a[8]*b[1]+a[7]*b[2]-a[6]*b[3]+a[5]*b[4]+a[4]*b[5]-a[3]*b[6]+a[2]*b[7]+a[1]*b[8]
	end,
	__concat = function(t,a,b)
t[1]=a[8]*b[1]+a[7]*b[2]-a[6]*b[3]+a[5]*b[4]+a[4]*b[5]-a[3]*b[6]+a[2]*b[7]+a[1]*b[8]
t[2]=a[8]*b[2]-a[6]*b[5]+a[5]*b[6]+a[2]*b[8]
t[3]=a[8]*b[3]-a[7]*b[5]+a[5]*b[7]+a[3]*b[8]
t[4]=a[8]*b[4]-a[7]*b[6]+a[6]*b[7]+a[4]*b[8]
t[5]=a[8]*b[5]+a[5]*b[8]
t[6]=a[8]*b[6]+a[6]*b[8]
t[7]=a[8]*b[7]+a[7]*b[8]
t[8]=a[8]*b[8]
	end,
	__len = function(t) t[1],t[2],t[3],t[4],t[5],t[6],t[7],t[8]=t[8],t[7],-t[6],t[5],t[4],-t[3],t[2],t[1] end,
	__unm = function(t) t[1],t[2],t[3],t[4],t[5],t[6],t[7],t[8]=-t[1],-t[2],-t[3],-t[4],-t[5],-t[6],-t[7],-t[8] end,
	rv = function(t) t[1],t[2],t[3],t[4],t[5],t[6],t[7],t[8]=t[1],t[2],t[3],t[4],-t[5],-t[6],-t[7],-t[8] end,
	iv = function(t) t[1],t[2],t[3],t[4],t[5],t[6],t[7],t[8]=t[1],-t[2],-t[3],-t[4],t[5],t[6],t[7],-t[8] end,
	cj = function(t) t[1],t[2],t[3],t[4],t[5],t[6],t[7],t[8]=t[1],-t[2],-t[3],-t[4],-t[5],-t[6],-t[7],t[8] end,
	lc = function(t,a,b)
t[1]=a[1]*b[1]+a[3]*b[3]+a[4]*b[4]-a[7]*b[7]
t[2]=a[1]*b[2]-a[3]*b[5]-a[4]*b[6]-a[7]*b[8]
t[3]=a[1]*b[3]-a[4]*b[7]
t[4]=a[1]*b[4]+a[3]*b[7]
t[5]=a[1]*b[5]+a[4]*b[8]
t[6]=a[1]*b[6]-a[3]*b[8]
t[7]=a[1]*b[7]
t[8]=a[1]*b[8]
	end,
	fd = function(t,a,b)
t[1]=a[1]*b[1]+a[3]*b[3]+a[4]*b[4]-a[7]*b[7]
t[2]=a[2]*b[1]+a[1]*b[2]+a[5]*b[3]+a[6]*b[4]-a[3]*b[5]-a[4]*b[6]-a[8]*b[7]-a[7]*b[8]
t[3]=a[3]*b[1]+a[1]*b[3]+a[7]*b[4]-a[4]*b[7]
t[4]=a[4]*b[1]-a[7]*b[3]+a[1]*b[4]+a[3]*b[7]
t[5]=a[5]*b[1]+a[8]*b[4]+a[1]*b[5]+a[4]*b[8]
t[6]=a[6]*b[1]-a[8]*b[3]+a[1]*b[6]-a[3]*b[8]
t[7]=a[7]*b[1]+a[1]*b[7]
t[8]=a[8]*b[1]+a[1]*b[8]
	end,
	ip = function(t,a,b)
t[1]=a[3]*b[3]+a[4]*b[4]-a[7]*b[7]
t[2]=a[5]*b[3]+a[6]*b[4]-a[3]*b[5]-a[4]*b[6]-a[8]*b[7]-a[7]*b[8]
t[3]=a[7]*b[4]-a[4]*b[7]
t[4]=-a[7]*b[3]+a[3]*b[7]
t[5]=a[8]*b[4]+a[4]*b[8]
t[6]=-a[8]*b[3]-a[3]*b[8]
	end,
	sp = function(t,a,b)
t[1]=a[1]*b[1]+a[3]*b[3]+a[4]*b[4]-a[7]*b[7]
	end
}

