return {
	__mul = function(t,a,b)
t[1]=a[1]*b[1]-a[9]*b[9]-a[10]*b[10]-a[11]*b[11]
t[6]=a[6]*b[1]+a[1]*b[6]+a[9]*b[7]+a[10]*b[8]-a[7]*b[9]-a[8]*b[10]-a[16]*b[11]-a[11]*b[16]
t[7]=a[7]*b[1]-a[9]*b[6]+a[1]*b[7]+a[11]*b[8]+a[6]*b[9]+a[16]*b[10]-a[8]*b[11]+a[10]*b[16]
t[8]=a[8]*b[1]-a[10]*b[6]-a[11]*b[7]+a[1]*b[8]-a[16]*b[9]+a[6]*b[10]+a[7]*b[11]-a[9]*b[16]
t[9]=a[9]*b[1]+a[1]*b[9]+a[11]*b[10]-a[10]*b[11]
t[10]=a[10]*b[1]-a[11]*b[9]+a[1]*b[10]+a[9]*b[11]
t[11]=a[11]*b[1]+a[10]*b[9]-a[9]*b[10]+a[1]*b[11]
t[16]=a[16]*b[1]+a[11]*b[6]-a[10]*b[7]+a[9]*b[8]+a[8]*b[9]-a[7]*b[10]+a[6]*b[11]+a[1]*b[16]
	end,
	__pow = function(t,a,b)
t[1]=a[1]*b[1]
t[6]=a[6]*b[1]+a[1]*b[6]
t[7]=a[7]*b[1]+a[1]*b[7]
t[8]=a[8]*b[1]+a[1]*b[8]
t[9]=a[9]*b[1]+a[1]*b[9]
t[10]=a[10]*b[1]+a[1]*b[10]
t[11]=a[11]*b[1]+a[1]*b[11]
t[16]=a[16]*b[1]+a[11]*b[6]-a[10]*b[7]+a[9]*b[8]+a[8]*b[9]-a[7]*b[10]+a[6]*b[11]+a[1]*b[16]
	end,
	__concat = function(t,a,b)
t[1]=a[16]*b[1]+a[11]*b[6]-a[10]*b[7]+a[9]*b[8]+a[8]*b[9]-a[7]*b[10]+a[6]*b[11]+a[1]*b[16]
t[6]=a[16]*b[6]+a[6]*b[16]
t[7]=a[16]*b[7]+a[7]*b[16]
t[8]=a[16]*b[8]+a[8]*b[16]
t[9]=a[16]*b[9]+a[9]*b[16]
t[10]=a[16]*b[10]+a[10]*b[16]
t[11]=a[16]*b[11]+a[11]*b[16]
t[16]=a[16]*b[16]
	end,
	__len = function(t) t[1],t[6],t[7],t[8],t[9],t[10],t[11],t[16]=t[16],t[11],-t[10],t[9],t[8],-t[7],t[6],t[1] end,
	__unm = function(t) t[1],t[6],t[7],t[8],t[9],t[10],t[11],t[16]=-t[1],-t[6],-t[7],-t[8],-t[9],-t[10],-t[11],-t[16] end,
	rv = function(t) t[1],t[6],t[7],t[8],t[9],t[10],t[11],t[16]=t[1],-t[6],-t[7],-t[8],-t[9],-t[10],-t[11],t[16] end,
	iv = function(t) t[1],t[6],t[7],t[8],t[9],t[10],t[11],t[16]=t[1],t[6],t[7],t[8],t[9],t[10],t[11],t[16] end,
	cj = function(t) t[1],t[6],t[7],t[8],t[9],t[10],t[11],t[16]=t[1],-t[6],-t[7],-t[8],-t[9],-t[10],-t[11],t[16] end,
	lc = function(t,a,b)
t[1]=a[1]*b[1]-a[9]*b[9]-a[10]*b[10]-a[11]*b[11]
t[6]=a[1]*b[6]-a[11]*b[16]
t[7]=a[1]*b[7]+a[10]*b[16]
t[8]=a[1]*b[8]-a[9]*b[16]
t[9]=a[1]*b[9]
t[10]=a[1]*b[10]
t[11]=a[1]*b[11]
t[16]=a[1]*b[16]
	end,
	fd = function(t,a,b)
t[1]=a[1]*b[1]-a[9]*b[9]-a[10]*b[10]-a[11]*b[11]
t[6]=a[6]*b[1]+a[1]*b[6]-a[16]*b[11]-a[11]*b[16]
t[7]=a[7]*b[1]+a[1]*b[7]+a[16]*b[10]+a[10]*b[16]
t[8]=a[8]*b[1]+a[1]*b[8]-a[16]*b[9]-a[9]*b[16]
t[9]=a[9]*b[1]+a[1]*b[9]
t[10]=a[10]*b[1]+a[1]*b[10]
t[11]=a[11]*b[1]+a[1]*b[11]
t[16]=a[16]*b[1]+a[1]*b[16]
	end,
	ip = function(t,a,b)
t[1]=-a[9]*b[9]-a[10]*b[10]-a[11]*b[11]
t[6]=-a[16]*b[11]-a[11]*b[16]
t[7]=a[16]*b[10]+a[10]*b[16]
t[8]=-a[16]*b[9]-a[9]*b[16]
	end,
	sp = function(t,a,b)
t[1]=a[1]*b[1]-a[9]*b[9]-a[10]*b[10]-a[11]*b[11]
	end
}
